File: ios/App/App/NetworkPrinterPlugin.swift
import Foundation
import Network
import Capacitor

@objc(NetworkPrinterPlugin)
public class NetworkPrinterPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NetworkPrinterPlugin"
    public let jsName = "NetworkPrinter"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "test", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "print", returnType: CAPPluginReturnPromise)
    ]

    @objc func test(_ call: CAPPluginCall) {
        ejecutar(call, datos: nil)
    }

    @objc func print(_ call: CAPPluginCall) {
        let texto = call.getString("texto") ?? ""
        let copias = min(max(call.getInt("copias") ?? 1, 1), 3)
        guard !texto.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("El ticket esta vacio.")
            return
        }
        ejecutar(call, datos: construirTrabajo(texto: texto, copias: copias))
    }

    private func ejecutar(_ call: CAPPluginCall, datos: Data?) {
        let ip = (call.getString("ip") ?? "").trimmingCharacters(in: .whitespaces)
        let puertoNumero = call.getInt("puerto") ?? 9100
        guard !ip.isEmpty,
              (1...65535).contains(puertoNumero),
              let puerto = NWEndpoint.Port(rawValue: UInt16(puertoNumero)) else {
            call.reject("La direccion IP o el puerto no son validos.")
            return
        }

        let conexion = NWConnection(host: NWEndpoint.Host(ip), port: puerto, using: .tcp)
        var finalizado = false
        let timeout = DispatchWorkItem {
            guard !finalizado else { return }
            finalizado = true
            conexion.cancel()
            call.reject("La conexion con la impresora agoto el tiempo de espera.")
        }

        conexion.stateUpdateHandler = { estado in
            switch estado {
            case .ready:
                timeout.cancel()
                if let datos = datos {
                    conexion.send(content: datos, completion: .contentProcessed { error in
                        guard !finalizado else { return }
                        finalizado = true
                        conexion.cancel()
                        if let error = error {
                            call.reject("No se pudo imprimir: \(error.localizedDescription)")
                        } else {
                            call.resolve(["mensaje": "Ticket enviado directamente desde el telefono."])
                        }
                    })
                } else {
                    guard !finalizado else { return }
                    finalizado = true
                    conexion.cancel()
                    call.resolve(["mensaje": "Conexion correcta con la impresora."])
                }
            case .failed(let error):
                timeout.cancel()
                guard !finalizado else { return }
                finalizado = true
                conexion.cancel()
                call.reject("No se pudo conectar con la impresora: \(error.localizedDescription)")
            default:
                break
            }
        }

        conexion.start(queue: DispatchQueue.global(qos: .userInitiated))
        DispatchQueue.global().asyncAfter(deadline: .now() + 5, execute: timeout)
    }

    private func construirTrabajo(texto: String, copias: Int) -> Data {
        var resultado = Data()
        let iniciar: [UInt8] = [0x1b, 0x40]
        let cortarStar: [UInt8] = [0x1b, 0x64, 0x02]
        let contenido = (texto + "\n").data(using: .ascii, allowLossyConversion: true) ?? Data()
        for _ in 0..<copias {
            resultado.append(contentsOf: iniciar)
            resultado.append(contenido)
            resultado.append(contentsOf: cortarStar)
        }
        return resultado
    }
}

File: android/app/src/main/java/com/hilos/posmobile/NetworkPrinterPlugin.java
package com.hilos.posmobile;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(name = "NetworkPrinter")
public class NetworkPrinterPlugin extends Plugin {
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    @PluginMethod
    public void test(PluginCall call) {
        executor.execute(() -> {
            try (Socket socket = conectar(call)) {
                JSObject resultado = new JSObject();
                resultado.put("mensaje", "Conexion correcta con la impresora.");
                call.resolve(resultado);
            } catch (Exception error) {
                call.reject(mensajeError(error), error);
            }
        });
    }

    @PluginMethod
    public void print(PluginCall call) {
        executor.execute(() -> {
            try (Socket socket = conectar(call)) {
                String texto = call.getString("texto", "");
                int copias = Math.min(Math.max(call.getInt("copias", 1), 1), 3);
                if (texto.trim().isEmpty()) throw new Exception("El ticket esta vacio.");

                OutputStream salida = socket.getOutputStream();
                salida.write(construirTrabajo(texto, copias));
                salida.flush();

                JSObject resultado = new JSObject();
                resultado.put("mensaje", "Ticket enviado directamente desde el telefono.");
                call.resolve(resultado);
            } catch (Exception error) {
                call.reject(mensajeError(error), error);
            }
        });
    }

    private Socket conectar(PluginCall call) throws Exception {
        String ip = call.getString("ip", "").trim();
        int puerto = call.getInt("puerto", 9100);
        if (!ip.matches("^(?:\\d{1,3}\\.){3}\\d{1,3}$")) {
            throw new Exception("Escribe una direccion IPv4 valida.");
        }
        Socket socket = new Socket();
        socket.connect(new InetSocketAddress(ip, puerto), 5000);
        socket.setSoTimeout(5000);
        return socket;
    }

    private byte[] construirTrabajo(String texto, int copias) throws Exception {
        ByteArrayOutputStream datos = new ByteArrayOutputStream();
        byte[] iniciar = new byte[] { 0x1b, 0x40 };
        byte[] cortarStar = new byte[] { 0x1b, 0x64, 0x02 };
        byte[] contenido = (texto + "\n").getBytes(StandardCharsets.US_ASCII);
        for (int i = 0; i < copias; i++) {
            datos.write(iniciar);
            datos.write(contenido);
            datos.write(cortarStar);
        }
        return datos.toByteArray();
    }

    private String mensajeError(Exception error) {
        String detalle = error.getMessage() == null ? "Error desconocido" : error.getMessage();
        return "No se pudo conectar con la impresora: " + detalle;
    }
}
